java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchDictionaryEncoder.java

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import org.apache.arrow.algorithm.search.VectorSearcher;
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.BitVectorHelper;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Dictionary encoder based on searching.
+ * @param <E> encoded vector type.
+ * @param <D> decoded vector type, which is also the dictionary type.
+ */
+public class SearchDictionaryEncoder<E extends BaseIntVector, D extends ValueVector> {
+
+  /**
+   * The dictionary for encoding/decoding.
+   * It must be sorted.
+   */
+  protected final D dictionary;
+
+  /**
+   * The criteria by which the dictionary is sorted.
+   */
+  protected final VectorValueComparator<D> comparator;
+
+  /**
+   * A flag indicating if null should be encoded.
+   */
+  protected final boolean encodeNull;
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary. It must be in sorted order.
+   * @param comparator the criteria for sorting.
+   */
+  public SearchDictionaryEncoder(D dictionary, VectorValueComparator<D> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary. It must be in sorted order.
+   * @param comparator the criteria for sorting.
+   * @param encodeNull if null should be encoded.
+   */
+  public SearchDictionaryEncoder(D dictionary, VectorValueComparator<D> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+  }
+
+  /**
+   * Encodes an input vector by binary search.
+   * So the algorithm takes O(nlogm) time, where n is the length of the input vector,
+   * and m is the length of the dictionary.
+   * @param input the input vector.
+   * @param output the output vector.
+   */
+  public void encode(D input, E output) {
+    for (int i = 0; i < input.getValueCount(); i++) {
+      if (!encodeNull && input.isNull(i)) {
+        BitVectorHelper.setValidityBit(output.getValidityBuffer(), i, 0);
+        continue;
+      }
+
+      int index = VectorSearcher.binarySearch(dictionary, comparator, input, i);
+      if (index == -1) {
+        throw new IllegalStateException("The data element is not found in the dictionary");
+      }
+      output.setWithPossibleTruncate(i, index);
+    }
+    output.setValueCount(input.getValueCount());
+  }
+
+  /**
+   * Decodes an input vector. The algorithm takes O(n) time,
+   * where n is the length of the input vector.
+   * @param input the input vector.
+   * @param output the output vector.
+   */
+  public void decode(E input, D output) {
+    for (int i = 0; i < input.getValueCount(); i++) {
+      if (!encodeNull && input.isNull(i)) {
```

Comment:
should this be here?  it seems like always want to set null if the input happens to be null?  

I wonder if it is quicker to copy the null buffer if there are any null values instead of doing this one by one.

Reply:
IMO, we do not always set null to tht output, if the input is null. We can treat null as a special element value, which takes 1 bit in the vilidity buffer, and 0 bit in the data buffer. Since it takes little space, we can use it to encode an extremely frequently used element to improve performance. 

I agree that it is much faster to copy the validity buffer directly, if null encoding is disabled.